### PR TITLE
Participating in Council Governance is Free for First Time Voters and Successful Closing

### PR DESCRIPTION
**Pull request by shamb0**
Participating in Governance should be Free for First Time Voters

Fixes: https://github.com/paritytech/substrate/issues/6904

**Summary of Code Changes**

| Title | Discp |
| --- | --- |
| Module | pallet-collective |
| Extrinsic functions | vote(), close() |

transaction fee will be waived for extrinsic calls vote() & close() for conditions described below ...

- In **vote()** api, first vote of the member in the motion is free, new logic got introduced to detect the member's first vote ( based on none status of ayes & nays ). and return value DispatchResultWithPostInfo is embedded with pay info ( Pays::No for first vote condition & Pays::No for all other conditions ).

- If the close operation completes successfully (with either approval or disapproval),the transaction fee will be waived. return value DispatchResultWithPostInfo is embedded with Pays::No under successful close conditions.

- Changes got verified with new Unit test function "motions_all_first_vote_free_works()"

### CODE REVIEW

File: frame/collective/src/lib.rs [hunk 1]
```diff
@@ -1441,6 +1458,186 @@ mod tests {
...
+			);
+
+			// Duplicate vote, expecting error with Pays::Yes ...
+			let vote_rval: DispatchResultWithPostInfo = Collective::vote(Origin::signed(2), hash.clone(), 0, true);
+			match vote_rval {
+				Ok(_) => {
+					println!( "@[{:#?}::{:#?}]::vote-fee() | Should not Occur",
+						file!(),
+						line!() );
+				},
```
Comment (shawntabrizi) [inline frame/collective/src/lib.rs:1493]: This should panic, forcing the test to fail.

File: frame/collective/src/lib.rs [hunk 2]
```diff
@@ -1441,6 +1458,186 @@ mod tests {
...
+				Err(err) => {
+					assert_eq!( err.post_info.pays_fee, Pays::Yes );
+				},
+			}
+
+			// Verify the events pool
+			assert_eq!(System::events(), vec![
```
Comment (shawntabrizi) [inline frame/collective/src/lib.rs:1566]: You def dont need this :)

File: frame/collective/src/lib.rs [hunk 3]
```diff
@@ -1441,6 +1458,186 @@ mod tests {
...
+				)
+			);
+
+			// trying to close the proposal, which is already closed. Expecting error "ProposalMissing" with Pays::Yes
+			let close_rval: DispatchResultWithPostInfo = Collective::close(Origin::signed(2), hash.clone(), 0, proposal_weight, proposal_len);
+			match close_rval {
+				Ok(_) => {
+					println!( "@[{:#?}::{:#?}]::vote-fee() | Should not Occur",
+						file!(),
+						line!() );
```
Comment (shawntabrizi) [inline frame/collective/src/lib.rs:1558]: Should be another panic situation, forcing the test to fail

File: frame/collective/src/lib.rs [hunk 4]
```diff
@@ -541,7 +544,18 @@ decl_module! {
...
 			Voting::<T, I>::insert(&proposal, voting);
 
-			Ok(Some(T::WeightInfo::vote(members.len() as u32)).into())
+			if is_account_voting_first_time {
+				Ok((
+					Some(T::WeightInfo::vote(members.len() as u32)),
+					Pays::No,
+				).into())
+			}
+			else {
```
Comment (shawntabrizi) [inline frame/collective/src/lib.rs:553]:
```suggestion
			} else {
```

**Review by shawntabrizi — APPROVED** [review]
Looks good to me!

File: frame/collective/src/lib.rs [hunk 5]
```diff
@@ -515,6 +515,9 @@ decl_module! {
 			let position_yes = voting.ayes.iter().position(|a| a == &who);
 			let position_no = voting.nays.iter().position(|a| a == &who);
 
+			// first vote of member in the motion
+			let is_account_voting_first_time = position_yes.is_none() && position_no.is_none();
```
Comment (shawntabrizi) [inline frame/collective/src/lib.rs:519]: This behavior should be mentioned in the docs above

File: frame/collective/src/lib.rs [hunk 6]
```diff
@@ -647,15 +663,16 @@ decl_module! {
 				Self::deposit_event(RawEvent::Closed(proposal_hash, yes_votes, no_votes));
 				let (proposal_weight, proposal_count) =
 					Self::do_approve_proposal(seats, voting, proposal_hash, proposal);
-				return Ok(Some(
-					T::WeightInfo::close_approved(len as u32, seats, proposal_count)
-						.saturating_add(proposal_weight)
+				return Ok((
+					Some(T::WeightInfo::close_approved(len as u32, seats, proposal_count).saturating_add(proposal_weight)),
+					Pays::No
```
Comment (shawntabrizi) [inline frame/collective/src/lib.rs:668]: This behavior should be mentioned in the docs above

File: frame/collective/src/lib.rs (hunk 1 above)
Comment (shamb0) [reply frame/collective/src/lib.rs:1492]: Thanks Shawn, replaced with suggested one.

File: frame/collective/src/lib.rs (hunk 3 above)
Comment (shamb0) [reply frame/collective/src/lib.rs:1557]: Thanks Shawn, replaced with suggested one.

File: frame/collective/src/lib.rs (hunk 5 above)
Comment (shamb0) [reply frame/collective/src/lib.rs:519]: Updated the function header Doc string ...

File: frame/collective/src/lib.rs (hunk 6 above)
Comment (shamb0) [reply frame/collective/src/lib.rs:667]: Updated the function header Doc string ...

File: frame/collective/src/lib.rs [hunk 7]
```diff
@@ -486,6 +486,8 @@ decl_module! {
 
 		/// Add an aye or nay vote for the sender to the given proposal.
 		///
+		/// And transaction fee is free for First time voters.
+		///
 		/// Requires the sender to be a member.
 		///
```
Comment (shawntabrizi) [inline frame/collective/src/lib.rs:492]:
```suggestion
		/// Requires the sender to be a member.
		///
		/// Transaction fees will be waived if the member is voting on any particular proposal
		/// for the first time and the call is successful.
```

File: frame/collective/src/lib.rs [hunk 8]
```diff
@@ -541,13 +546,25 @@ decl_module! {
...
+					Pays::Yes,
+				).into())
+			}
 		}
 
 		/// Close a vote that is either approved, disapproved or whose voting period has ended.
 		///
 		/// May be called by any signed account in order to finish voting and close the proposal.
 		///
+		/// And transaction fee is free for successful close.
```
Comment (shawntabrizi) [inline frame/collective/src/lib.rs:566]:
```suggestion
		/// If the close operation completes successfully (with either approval or disapproval),
		/// the transaction fee will be waived.
```

Comment (shawntabrizi) [inline frame/collective/src/lib.rs:566]:
```suggestion
		/// If the close operation completes successfully (with either approval or disapproval),
		/// the transaction fee will be waived.
```

File: frame/collective/src/lib.rs [hunk 9]
```diff
@@ -647,15 +668,17 @@ decl_module! {
 				Self::deposit_event(RawEvent::Closed(proposal_hash, yes_votes, no_votes));
 				let (proposal_weight, proposal_count) =
 					Self::do_approve_proposal(seats, voting, proposal_hash, proposal);
-				return Ok(Some(
-					T::WeightInfo::close_approved(len as u32, seats, proposal_count)
-						.saturating_add(proposal_weight)
+				return Ok((
+					Some(T::WeightInfo::close_approved(len as u32, seats, proposal_count)
+					.saturating_add(proposal_weight)),
+					Pays::No
```
Comment (shawntabrizi) [inline frame/collective/src/lib.rs:674]:
```suggestion
					Pays::No,
```

Small nit pick detail, we like to put a comma at the end of newline separated lists.

https://users.rust-lang.org/t/trailing-commas/13993

Few more instances of this in other spots in this PR

**Comment by shawntabrizi** [comment]
Looks ready to merge once we have another review and an audit. Audit team is made aware and will look next week.

File: frame/collective/src/lib.rs (hunk 9 above)
Comment (shamb0) [reply frame/collective/src/lib.rs]: Thanks Shawn, will review again and fix it

**Comment by shamb0** [comment]
Fine Shawn, And Trailing Commas are added ...

File: frame/collective/src/lib.rs [hunk 10]
```diff
@@ -1441,6 +1464,145 @@ mod tests {
...
+					Box::new(proposal.clone()),
+					proposal_len,
+				)
+			);
+			assert_eq!(
+				Collective::voting(&hash),
+				Some(Votes { index: 0, threshold: 2, ayes: vec![1], nays: vec![], end })
+			);
+
+			// For the motion, acc 2's first vote, expecting Ok with Pays::No ...
```
Comment (kianenigma) [inline frame/collective/src/lib.rs:1487]:
```suggestion
			// For the motion, acc 2's first vote, expecting Ok with Pays::No.
```

File: frame/collective/src/lib.rs [hunk 11]
```diff
@@ -1441,6 +1464,145 @@ mod tests {
...
+					PostDispatchInfo {
+						actual_weight: Some(
+							207711000,
+						),
+						pays_fee: Pays::No,
+					}
+				)
+			);
+
+			// Duplicate vote, expecting error with Pays::Yes ...
```
Comment (kianenigma) [inline frame/collective/src/lib.rs:1500]:
```suggestion
			// Duplicate vote, expecting error with Pays::Yes.
```

File: frame/collective/src/lib.rs [hunk 12]
```diff
@@ -1441,6 +1464,145 @@ mod tests {
...
+					// Forced Panic, Invalid case, should not occur.
+					panic!(
+						"Invalid State::Duplicate vote, error expected - invalid rval {:#?}",
+						rval,
+					);
+				},
+				Err(err) => {
+					assert_eq!(err.post_info.pays_fee, Pays::Yes);
+				},
+			}
```
Comment (kianenigma) [inline frame/collective/src/lib.rs:1518]:
```suggestion
			assert_eq!(vote_rval.unwrap_err().post_info.pays_fee, Pays::Yes);
```

File: frame/collective/src/lib.rs [hunk 13]
```diff
@@ -1441,6 +1464,145 @@ mod tests {
...
+
+			// Modifying vote, expecting ok with Pays::Yes ...
```
Comment (kianenigma) [inline frame/collective/src/lib.rs:1520]:
```suggestion
			// Modifying vote, expecting ok with Pays::Yes.
```

File: frame/collective/src/lib.rs [hunk 14]
```diff
@@ -1441,6 +1464,145 @@ mod tests {
...
+					PostDispatchInfo {
+						actual_weight: Some(
+							207711000,
+						),
+						pays_fee: Pays::Yes,
+					}
+				)
+			);
+
+			// For the motion, acc 3's first vote, expecting Ok with Pays::No ...
```
Comment (kianenigma) [inline frame/collective/src/lib.rs:1533]:
```suggestion
			// For the motion, acc 3's first vote, expecting Ok with Pays::No.
```

File: frame/collective/src/lib.rs [hunk 15]
```diff
@@ -1441,6 +1464,145 @@ mod tests {
...
+					PostDispatchInfo {
+						actual_weight: Some(
+							207711000,
+						),
+						pays_fee: Pays::No,
+					}
+				)
+			);
+
+			// acc 3 modify the vote, expecting Ok with Pays::Yes ...
```
Comment (kianenigma) [inline frame/collective/src/lib.rs:1546]:
```suggestion
			// acc 3 modify the vote, expecting Ok with Pays::Yes.
```

File: frame/collective/src/lib.rs [hunk 16]
```diff
@@ -1441,6 +1464,145 @@ mod tests {
...
+					// Forced Panic, Invalid case, should not occur.
+					panic!(
+						"Invalid State::trying to close the invalid proposal, invalid rval {:#?}",
+						rval,
+					);
+				},
+				Err(err) => {
+					assert_eq!(err.post_info.pays_fee, Pays::Yes);
+				},
+			}
```
Comment (kianenigma) [inline frame/collective/src/lib.rs:1602]:
```suggestion
			assert_eq!(vote_rval.unwrap_err().post_info.pays_fee, Pays::Yes);
```

**Review by kianenigma — APPROVED** [review]
Could use my comments, but all in all looks good.

**Comment by shamb0** [comment]
Thanks @kianenigma for review & comments, will make sure not repeat the minor mistakes again :-)

**Comment by redzsina** [comment]
We are done with our review for this PR - please refer to issue 30 in the internal tracker.

**Review by thiolliere — APPROVED** [review]
I don't have opinion on whether those calls should be free, but code looks good and logic can't be abused

**Comment by shawntabrizi** [comment]
bot merge

**Pull request merged**